docs/design/resource-qos.md

```diff
@@ -0,0 +1,224 @@
+For example, the scheduler would not place 5 containers each with a 1GB memory limit onto a machine with 4GB of memory.
+
+This had the drawback that if users over-estimate the resource needs of their containers, then cluster utilization will be poor, since the unused resources are stranded and can't be used by other containers.
+On the other hand, if they under-estimate the resource needs (specifying no limits is a special case of under-estimating), then the machine could become overcommitted, leading to out-of-memory or CPU starvation of any container on the node, even the ones that specified accurate resource limits.
+
+This document describes the way Kubernetes provides different levels of Quality of Service to containers depending on what they *request*.
+Containers that need to stay up reliably can request guaranteed resources, while containers with less stringent requirements can use resources with weaker or no guarantee.
+Specifically, for each resource, containers specify a request, which is the amount of that resource that the system will guarantee to the container, and a limit which is the maximum amount that the system will allow the container to use.
+When request == limit, the resources are guranteed, and when request < limit, the container is guaranteed the request but can opportunistically scavenge the difference between request and limit if they are not being used by other containers.
+This allows Kubernetes to oversubscribe nodes, which increases utilization, while at the same time maintaining resource guarantees for the containers that need guarantees.
+Borg increased utilization by about 20% when it started allowing use of such non-guaranteed resources, and we hope to see similar improvements in Kubernetes.
+
+## Requests and Limits
+
+For each resource, containers can specify a resource request and limit, 0 <= request <= limit <= Infinity.
+If the container is successfully scheduled, the container is guaranteed the amount of resource requested.
+Scheduling is based on `requests` and not `limits`.
+The container will not be allowed to exceed the specified limit.
+How the request and limit are enforced depends on whether the resource is [compressible or incompressible](resources.md).
+
+### Compressible Resource Guarantees
+
+- For now, we are only supporting CPU.
+- Containers are guaranteed to get the amount of CPU they request, they may or may not get additional CPU time (depending on the other jobs running).
+- Excess CPU resources will be distributed based on the amount of CPU requested. For example, suppose container A requests for 60% of the CPU, and container B requests for 30% of the CPU. Suppose that both containers are trying to use as much CPU as they can. Then the extra 10% of CPU will be distributed to A and B in a 2:1 ratio (implementation discussed in later sections).
+- Containers will be throttled if they exceed their limit. If limit is unspecified, then the containers can use excess CPU when available.
```

Comment:
I was never aware of such requirements :) 
I added a note on CPU limits. 
Thanks for pointing this out.